source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -0,0 +1,221 @@
+                         *decoder_callbacks_, data, end_stream);
+}
+
+Http::FilterHeadersStatus DecompressorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                            bool end_stream) {
+  // Headers only response, continue.
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ENVOY_STREAM_LOG(debug, "DecompressorFilter::encodeHeaders: {}", *encoder_callbacks_, headers);
+
+  return maybeInitDecompress(config_->responseDirectionConfig(), response_decompressor_,
+                             encoder_callbacks_, *encoder_callbacks_, headers);
+}
+
+Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  return maybeDecompress(config_->responseDirectionConfig(), response_decompressor_.get(),
+                         *encoder_callbacks_, data, end_stream);
+}
+
+Http::FilterHeadersStatus DecompressorFilter::maybeInitDecompress(
+    DecompressorFilterConfig::DirectionConfig& direction_config,
+    Compression::Decompressor::DecompressorPtr& decompressor,
+    absl::variant<Http::StreamDecoderFilterCallbacks*, Http::StreamEncoderFilterCallbacks*>
+        decoder_or_encoder_callbacks,
+    Http::StreamFilterCallbacks& callbacks, Http::RequestOrResponseHeaderMap& headers) {
+  if (direction_config.decompressionEnabled() && !hasCacheControlNoTransform(headers) &&
+      contentEncodingMatches(headers)) {
+    direction_config.stats().decompressed_.inc();
+    decompressor = config_->makeDecompressor();
+    // remove the compression scheme's content encoding because the filter is going to decompress.
+    removeContentEncoding(headers);
+
+    // Per the filter documentation, only buffer if there is a Content-Length header present, and
+    // the filter is configured to buffer.
+    if (headers.removeContentLength() > 0 && direction_config.maxBufferedBytes()) {
+      direction_config.setBuffering(true);
+      if (absl::holds_alternative<Http::StreamDecoderFilterCallbacks*>(
+              decoder_or_encoder_callbacks)) {
+        absl::get<Http::StreamDecoderFilterCallbacks*>(decoder_or_encoder_callbacks)
+            ->setDecoderBufferLimit(direction_config.maxBufferedBytes().value());
+      } else {
+        absl::get<Http::StreamEncoderFilterCallbacks*>(decoder_or_encoder_callbacks)
+            ->setEncoderBufferLimit(direction_config.maxBufferedBytes().value());
+      }
+      direction_config.setHeaders(&headers);
+      // Note that the log will print the updated headers. The incoming headers can be seen from the
+      // log in decode/encodeHeaders.
+      ENVOY_STREAM_LOG(debug, "do decompress (without buffering) {}: {}", callbacks,
+                       direction_config.logString(), headers);
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+    // FIX ME: chunked should not be added if we buffer, right?
+    sanitizeTransferEncoding(headers);
```

Comment:
That is my understanding too: if `chunked` is added then `Content-Length` has to be omitted.